wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceHolder.java

```diff
@@ -85,35 +87,99 @@ public UsersWorkspace getWorkspace() {
-        final Link link = newDto(Link.class).withMethod("PUT").withHref(href);
+        final Link link = newDto(Link.class).withMethod(POST).withHref(href);
 
         try {
-            httpJsonRequestFactory.fromLink(link)
-                                  .setBody(DtoConverter.asDto(workspace.getConfig()))
-                                  .request();
+            httpJsonRequestFactory.fromLink(link).setBody(asDto(project)).request();
         } catch (IOException | ApiException e) {
             throw new ServerException(e.getMessage());
         }
 
-        // sync local projects
-        projects.stream()
-                .filter(project -> !project.isSynced())
-                .forEach(RegisteredProject::setSync);
+        if (!project.isSynced()) {
+            project.setSync();
+        }
+    }
+
+    /**
+     * Updates project on WS-master side.
+     *
+     * @param project
+     *         project to update
+     * @throws ServerException
+     */
+    void updateProject(RegisteredProject project) throws ServerException {
+        if (project.isDetected()) {
+            return;
+        }
+
+        // TODO workspace.addProject(project); but replace
+        workspace.updateProject(project);
+
+        final String href = UriBuilder.fromUri(apiEndpoint)
+                                      .path(WorkspaceService.class)
+                                      .path(WorkspaceService.class, "updateProject")
+                                      .build(workspace.getId()).toString();
+        final Link link = newDto(Link.class).withMethod(PUT).withHref(href);
+
+        try {
+            httpJsonRequestFactory.fromLink(link).setBody(asDto(project)).request();
+        } catch (IOException | ApiException e) {
+            throw new ServerException(e.getMessage());
+        }
+
+        if (!project.isSynced()) {
+            project.setSync();
+        }
+    }
+
+    /**
+     * Removes projects on WS-master side.
+     *
+     * @param projects
+     *         projects to remove
+     * @throws ServerException
+     */
+    void removeProjects(Collection<RegisteredProject> projects) throws ServerException {
+        for (RegisteredProject project : projects) {
+            removeProject(project);
+        }
+    }
+
+    private void removeProject(RegisteredProject project) throws ServerException {
+        if (project.isDetected()) {
+            return;
+        }
+
+        workspace.removeProject(project);
+
+        final String href = UriBuilder.fromUri(apiEndpoint)
+                                      .path(WorkspaceService.class)
+                                      .path(WorkspaceService.class, "deleteProject")
+                                      .build(workspace.getId(), project.getName()).toString();
+        final Link link = newDto(Link.class).withMethod(DELETE).withHref(href);
+
+        try {
+            httpJsonRequestFactory.fromLink(link).request();
```

Comment:
WDYT about using of 
```
httpJsonRequestFactory.fromUrl(href)
                      .useDeleteMethod()
                      .request();
``` 
instead of creating new  DTO Link?